File: 프로그래머스/3/43164. 여행경로/여행경로.py
def solution(tickets):
    # 1. 그래프 만들기
    graph = {}
    for a,b in tickets:
        if a not in graph:
            graph[a] = []
        graph[a].append(b)
    
    # 2. 사전 순으로 정렬
    sorted_graph = dict(sorted(graph.items(), key=lambda x: (x[0])))
    for key in graph:
        graph[key].sort()
        
    """for key, value in sorted_graph.items():
        print(key, value)"""
        
    route = []  # 최종 경로

    def dfs(airport):
        # 현재 공항에서 가능한 목적지가 없을 때
        if airport not in sorted_graph or not sorted_graph[airport]:
            route.append(airport)
            return
        while sorted_graph[airport]:
            next_airport = sorted_graph[airport].pop(0)  # 작은 알파벳 우선
            dfs(next_airport)
        route.append(airport)
    
    dfs("ICN")
    return route[::-1]  # 역순 반환 (DFS 종료 시점 순서이기 때문)

File: 프로그래머스/3/43164. 여행경로/여행경로.js
function solution(tickets) {
    var answer = [];
    
    // 0. 티켓 정렬
    tickets.sort((a,b)=>{
        if(a[0]===b[0]){
            return a[1].localeCompare(b[1]);
        }
        return a[0].localeCompare(b[0]);
    });
    //console.log(tickets)
    
    // 1. 연결리스트 만들기
    let graph = [];
    for(const [a,b] of tickets){
        if (!(a in graph)){
            graph[a] = []
        }
        graph[a].push(b)
    }
    console.log(graph)
    
    // 3. dfs 함수
    let route = [];
    const dfs = (airport) => {
        if (!(airport in graph)||graph[airport].length === 0){
            route.push(airport);
            return
        }
        while (graph[airport].length > 0){
            next_airport = graph[airport].shift();
            dfs(next_airport)
        }
        route.push(airport);
    }
    dfs("ICN");
    answer = route.slice().reverse();
    return answer;
}